devito/passes/clusters/implicit.py

```diff
@@ -0,0 +1,133 @@
+    """
+    Bind the free thickness symbols defined by MultiSubDimensions to
+    suitable values.
+
+    Examples
+    --------
+    Given:
+
+        Cluster([Eq(f[t1, xi_n, yi_n], f[t0, xi_n, yi_n] + 1)])
+
+    where `xi_n` and `yi_n` are MultiSubDimensions, generate:
+
+        Cluster([Eq(xi_n_ltkn, xi_n_m[n])
+                 Eq(xi_n_rtkn, xi_n_M[n])
+                 Eq(yi_n_ltkn, yi_n_m[n])
+                 Eq(yi_n_rtkn, yi_n_M[n])])
+        Cluster([Eq(f[t1, xi_n, yi_n], f[t0, xi_n, yi_n] + 1)])
+    """
+
+    def callback(self, clusters, prefix):
+        try:
+            dd = prefix[-1].dim
+        except IndexError:
+            dd = None
+
+        # The non-MultiSubDimension closest to a MultiSubDimension triggers
+        # the pass. For example, `t` in an `t, xi_n, yi_n` iteration space
+        if msdim(dd):
+            return clusters
+
+        seen = set()
+        tip = None
+
+        processed = []
+        for c in clusters:
+            try:
+                idx = len(prefix)
+                dd = c.ispace[idx].dim
+                d = msdim(dd)
+            except IndexError:
+                d = None
+            if d is None:
+                processed.append(c)
+                continue
+
+            idx = c.ispace.index(dd)
+            ispace0 = c.ispace[:idx]
+            ispace1 = c.ispace[idx:]
+
+            # The implicit expressions introduced by the MultiSubDomain
+            exprs = d.msd._implicit_exprs
+
+            # The implicit Dimensions and iterators induced by the MultiSubDomain
+            dims = filter_ordered(retrieve_dimensions(exprs, deep=True))
+            idims = tuple(i for i in dims if not i.is_SubIterator)
+            intervals = [Interval(i, 0, 0) for i in idims]
+            sub_iterators = {i.root: i for i in dims if i.is_SubIterator}
+
+            # The local IterationSpace of the implicit Dimensions, if any
+            relations = (ispace0.itdimensions + idims,
+                         idims + ispace1.itdimensions)
+            ispaceN = IterationSpace(
+                IntervalGroup(intervals, relations=relations),
+                sub_iterators
+            )
+
+            ispace = IterationSpace.union(ispace0, ispaceN)
+            if len(ispaceN) == 0:
+                # Special case: we can factorize the thickness assignments
+                # once and for all at the top of the current IterationInterval
+                if ispaceN not in seen:
+                    processed.insert(0, Cluster(exprs, ispace))
+                    seen.add(ispaceN)
+            else:
+                nxt = self._make_tip(c, ispaceN)
+                if tip is None or tip != nxt:
+                    processed.append(c.rebuild(exprs=exprs, ispace=ispace))
+                    tip = nxt
+                else:
+                    # Special case: we're reusing the previously constructed ispaceN
+                    pass
+
+            ispace = IterationSpace.union(c.ispace, ispaceN)
```

Comment:
not really

The former ispace is for the thickness assignments
The latter ispace is for the reconstructed cluster

While the former may occasionally be skipped (optimization to avoid redundant assignments), the latter is compulsory